### PR TITLE
Feature/undo redo

### PR DESCRIPTION
**Pull request by Bushuo**
Ref: #306 

## Basics

- [x] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildserver is happy.
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

**Comment by buenaflor** [comment]
Nice! how does the performance look like when adding x numbers of objects to the history? But looks good to me 👍

**Review by buenaflor — APPROVED** [review]

**Comment by Bushuo** [comment]
> Nice! how does the performance look like when adding x numbers of objects to the history? But looks good to me 👍

Good question 😄 
For each action, only the payload of the action is added to the history. 
I think, therefore the runtime complexity of the undo/redo step grows in the number of history steps linearly and it does not grow in the number of elements on the screen.

I thought this was a better tradeoff.
The other proposed solution would have been copying the whole state for each action, which is O(n) in the number of elements on the map and is not very memory efficient.

File: frontend/src/features/undo_redo/store/MapHistoryStore.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+function handleUpdateObjectAction(
+  state: MapState,
+  action: ObjectUpdatePositionAction | ObjectUpdateTransformAction,
+): MapState {
+  return {
+    ...state,
+    layers: {
+      ...state.layers,
+      ...action.payload.reduce(
```
Comment (absurd-turtle) [inline frontend/src/features/undo_redo/store/MapHistoryStore.ts:133]: Can you extract the anonymous functions? I find this quite hard to read.

**Review by absurd-turtle — CHANGES_REQUESTED** [review]
Looks pretty good to me!

I think it would make sense to create some unit tests for the functions in MapHistoryStore.ts

**Comment by Bushuo** [comment]
> Looks pretty good to me!
> 
> I think it would make sense to create some unit tests for the functions in MapHistoryStore.ts

@absurd-turtle I added unit tests and made the code a little more readable.
Could you please check again?

File: frontend/src/features/map_planning/components/BaseStage.tsx [hunk 2]
```diff
@@ -156,29 +166,105 @@ export const BaseStage = ({
...
     <div className="h-screen w-screen overflow-hidden">
+      <div className="absolute z-10 flex h-10 items-center gap-2 pl-2 pt-12">
+        <SimpleButton
+          className="w-32"
+          onClick={() =>
+            dispatch({
+              type: 'OBJECT_ADD',
+              payload: {
+                index: 'plant',
+                id: Math.random().toString(36).slice(2, 9),
```
Comment (absurd-turtle) [inline frontend/src/features/map_planning/components/BaseStage.tsx:200]: Is this save enough in terms of uniqueness. Are there any consequences when two object have the same id?

File: frontend/src/features/map_planning/routes/Map.tsx [hunk 3]
```diff
@@ -11,39 +12,38 @@ import { Circle, Rect } from 'react-konva';
...
+            {...o}
+            key={o.id}
+            fill="blue"
+            draggable={true}
+            shadowBlur={5}
+            onClick={(e) => {
+              addToTransformer(e.target as Shape<ShapeConfig>);
+            }}
+            onDragStart={(e) => {
+              // sometimes the click event is not fired, so we have to add the object to the transformer here
```
Comment (absurd-turtle) [inline frontend/src/features/map_planning/routes/Map.tsx:41]: Is the object added twice when both events fire?

**Review by absurd-turtle — APPROVED** [review]
Good job adding the tests! :)
There are two minor points I want to discuss but it looks good otherwise and is ready to be merged then.

File: frontend/src/features/map_planning/components/BaseStage.tsx (hunk 2 above)
Comment (Bushuo) [reply frontend/src/features/map_planning/components/BaseStage.tsx:200]: @absurd-turtle This is example code, that shows how an object should be added to the store. 
I will add a comment that this is still a TODO but it should be handled in another issue.

Comment (absurd-turtle) [reply frontend/src/features/map_planning/components/BaseStage.tsx:200]: Ah ok, that makes sense

File: frontend/src/features/map_planning/routes/Map.tsx (hunk 3 above)
Comment (Bushuo) [reply frontend/src/features/map_planning/routes/Map.tsx:41]: No, the reference would be equal and only the node from the first event is added to the transformer. The second can not pass the check in line 22
```ts
if (!nodes.includes(node)) {
  transformer?.nodes([node]);
}
```

Comment (absurd-turtle) [reply frontend/src/features/map_planning/routes/Map.tsx:41]: makes sense

**Review by markus2330 — APPROVED** [review]
Great job, looking forward to get this merged but it needs rebase.

**Comment by markus2330** [comment]
@Bushuo There is a CI problem now.

**Comment by Bushuo** [comment]
jenkins build please

**Comment by Bushuo** [comment]
@markus2330 weird, I don't have any problems locally even when executing the same script with clean `node_modules`. Let's see if the problem persists.

**Comment by markus2330** [comment]
@Bushuo you removed [frontend/package-lock.json](https://github.com/ElektraInitiative/PermaplanT/pull/308/files#diff-4a2d9aa3e849b134993936ca81b83fb139edd2b0218077ab0f403b8c4803c62a) in this PR, is this on purpose or is this maybe the cause of the problem?

**Comment by Bushuo** [comment]
> @Bushuo you removed [frontend/package-lock.json](https://github.com/ElektraInitiative/PermaplanT/pull/308/files#diff-4a2d9aa3e849b134993936ca81b83fb139edd2b0218077ab0f403b8c4803c62a) in this PR, is this on purpose or is this maybe the cause of the problem?

I'm pretty sure that I did not remove it, but I had to resolve merge conflicts there. For this is just executed `npm i` which automatically resolves the conflicts in `package-lock.json`.

I think that this is a problem with the CI. The previous build step `check-frontend` inside the same execution of the pipeline ran `npm ci` without failure.

**Pull request merged**